cdap-master/src/main/java/co/cask/cdap/data/runtime/main/StartupChecker.java

```diff
@@ -0,0 +1,377 @@
+                                   Constants.Stream.CONTAINER_INSTANCES,
+                                   Constants.Stream.MAX_INSTANCES))
+      .add(new ServiceResourceKeys(Constants.Service.METRICS,
+                                   Constants.Metrics.MEMORY_MB,
+                                   Constants.Metrics.NUM_CORES,
+                                   Constants.Metrics.NUM_INSTANCES,
+                                   Constants.Metrics.MAX_INSTANCES))
+      .add(new ServiceResourceKeys(Constants.Service.METRICS_PROCESSOR,
+                                   Constants.MetricsProcessor.MEMORY_MB,
+                                   Constants.MetricsProcessor.NUM_CORES,
+                                   Constants.MetricsProcessor.NUM_INSTANCES,
+                                   Constants.MetricsProcessor.MAX_INSTANCES))
+      .add(new ServiceResourceKeys(Constants.Service.LOGSAVER,
+                                   Constants.LogSaver.MEMORY_MB,
+                                   Constants.LogSaver.NUM_CORES,
+                                   Constants.LogSaver.NUM_INSTANCES,
+                                   Constants.LogSaver.MAX_INSTANCES));
+    if (cConf.getBoolean(Constants.Explore.EXPLORE_ENABLED)) {
+      builder.add(new ServiceResourceKeys(Constants.Service.EXPLORE_HTTP_USER_SERVICE,
+                                          Constants.Explore.CONTAINER_MEMORY_MB,
+                                          Constants.Explore.CONTAINER_VIRTUAL_CORES,
+                                          Constants.Explore.CONTAINER_INSTANCES,
+                                          Constants.Explore.MAX_INSTANCES));
+    }
+    this.systemServicesResourceKeys = builder.build();
+  }
+
+  /**
+   * Run sanity checks to see if there are any glaring problems that will prevent master from starting up.
+   * These are the types of things that are not likely to be resolved without manual intervention, such as incorrect
+   * directory permissions, wrong hive dependencies, and underlying infrastructure like HDFS being unavailable.
+   *
+   * Other types of errors that may eventually resolve themselves, such as failure to connect to transaction service,
+   * are handled in a separate method.
+   *
+   * This method is a no-op if {@link Constants#STARTUP_CHECKS_ENABLED} is false.
+   *
+   * @return whether the sanity checks succeeded or not
+   */
+  public boolean runSanityChecks() {
+    boolean noErrors = true;
+
+    ConfigurationLogger.logImportantConfig(cConf);
+    if (cConf.getBoolean(Constants.STARTUP_CHECKS_ENABLED)) {
+      noErrors = checkConfig();
+      noErrors = checkHDFS() && noErrors;
+      noErrors = checkYarn() && noErrors;
+      noErrors = checkHBase() && noErrors;
+    }
+
+    return noErrors;
+  }
+
+  // TODO: (CDAP-4517) add more checks, like zookeeper settings.
+  // check cdap config settings. For example, memory.mb settings should be numbers and positive.
+  private boolean checkConfig() {
```

Comment:
Or, instead of loading all checkers, you can load a configuration containing the list of checker classes to run, which basically can replace the `STARTUP_CHECKS_ENABLED` option (if the config is an empty string, then there is nothing to check).